lib/common/xml.c

```diff
@@ -1754,7 +1754,7 @@ log_data_element(int log_level, const char *file, const char *function,
     }
 
     if (prefix == NULL) {
-        prefix = "";
+        prefix = "  ";
```

Comment:
This is really a guess, so I'd appreciate some insight here. Are the various prefixes and their meanings documented anywhere? There are a bunch of `+`, `-`, and `~` symbol combinations scattered throughout the patchset/xml code.

Below here, we set the element at index 1 to either `'+'` or `'-'` if particular flags are set, after defaulting `prefix` to an empty string -- so that's an out-of-bounds access. This is an attempt to fix that issue. But... we're only modifying the second element below. What's supposed to be in the first element?

For example, if the prefix is some arbitrary string, it doesn't make sense to overwrite the second character. And if the first character is a plus, I'm not sure if it makes sense for the second character to be a minus.

Reply:
> This is really a guess, so I'd appreciate some insight here. Are the various prefixes and their meanings documented anywhere? There are a bunch of `+`, `-`, and `~` symbol combinations scattered throughout the patchset/xml code.

Undocumented, but I believe `--` indicates something removed from the original CIB, `++` indicates something added or replaced, and `+~` indicates something moved. `- ` or `+ ` identifies the original or patched CIB somehow.
 
> Below here, we set the element at index 1 to either `'+'` or `'-'` if particular flags are set, after defaulting `prefix` to an empty string -- so that's an out-of-bounds access. This is an attempt to fix that issue. But... we're only modifying the second element below. What's supposed to be in the first element?

I suspect prefix never being empty is the only thing avoiding a segfault. We could CRM_CHECK() it.

> For example, if the prefix is some arbitrary string, it doesn't make sense to overwrite the second character. And if the first character is a plus, I'm not sure if it makes sense for the second character to be a minus.

I suspect `--`, `++`, and `+~` are the only possible two-character combinations in practice.

Reply:
> Undocumented, but I believe -- indicates something removed from the original CIB, `++` indicates something added or replaced, and `+~` indicates something moved. `-` or `+` identifies the original or patched CIB somehow.

That sounds right... it looks like callers often pass in `"+ "` or `"- "` with the respective `xml_log_option_diff_*` option. Then if `xml_log_option_dirty_add` is not set, `log_data_element` overwrites the space with the second `'+'` or `'-'` character. If `xml_log_option_dirty_add` is set, then it calls `log_xml_changes()`, which sets the first two characters to `"++"`, `"--"`, or `"+~"`.

It might be worth redoing it a little bit more clearly but I'm not sure when I'm gonna want to do that.

> I suspect prefix never being empty is the only thing avoiding a segfault. We could CRM_CHECK() it.

That'll work.

> I suspect `--`, `++`, and `+~` are the only possible two-character combinations in practice.

The only meaningful ones at least. For example, `do_crm_log_xml()` passes an arbitrary string as the prefix. That's part of what I was worried about -- overwriting the first and/or second character. I think we're safe in that case because the options flag group is hard-coded as `xml_log_option_formatted`, so we won't have any of the diff or dirty_add flags set.